### PR TITLE
Evaluating all models

### PR DESCRIPTION
**Pull request by jdiaz4302**
This PR is the evaluation version of the PR which extended training to all model types.

# Overview

All of the same calculations are done here (daily accuracy, error of ice on date, error of ice off date, and error of ice duration) for every trained model except the massive transformers (N = 25), which predict so poorly that they break the functions (due to not exceeding a predicted ice probability of 50% ever)

New additions beyond looping over models:
- Plotting all validation loss curves together. This was originally done in the training notebook. Note that number of epochs are _not_ synonymous with training time.
- Averaging date-based RMSEs to use as a final determination of improvement over process-based predictions.
- Switched from reporting Pearson R to R-squared in plots of observations versus predictions

# Requested review

Identification of bugs or erroneous implementation. Everything seems to be working as expected for many iterations at this point, but, similar to the positional encoding in the last PR, there could always be something incorrect that sneaks by.

### CODE REVIEW

File: 03_eval/src/A_EvaluationMetrics.md [hunk 1]
```diff
@@ -47,13 +46,39 @@ valid_predictions_fpath = train_out_dir + 'limitted_lstm_valid_preds.npy'
...
+              'massive transformer':'#e6ab02',}
 ```
 
 ### Outputs
 
 ```python
-eval_metrics_fpath = '../out/eval_metrics.npz'
+# Fragile: prepare to change if new results; this is a little qualitative atm so not automated
+second_best_eval_metrics_fpath = '../out/avg_lstm_eval_metrics_3_.npz'
+best_eval_metrics_fpath = '../out/massive_lstm_eval_metrics_1_.npz'
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.md:81]: These get identified as best models (avg lstm seed 3, massive lstm seed 1)

File: 03_eval/src/A_EvaluationMetrics.md [hunk 2]
```diff
@@ -79,8 +104,46 @@ valid_variables = valid_data['features']
 ```
 
 ```python
-train_predictions = np.load(train_predictions_fpath)
-valid_predictions = np.load(valid_predictions_fpath)
+# Objects to store data in
+all_train_predictions = np.zeros([n_compare, train_n, seq_len, 1])
+all_valid_predictions = np.zeros([n_compare, valid_n, seq_len, 1])
+all_valid_loss_values = np.zeros([n_compare, n_epochs])
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.md:110]: Training all the models on Tallgrass default gpu interactive instance which does not have matplotlib, so visualizing these curves here

File: 03_eval/src/A_EvaluationMetrics.md [hunk 3]
```diff
@@ -266,149 +457,292 @@ def calc_ice_duration(off, on):
...
+# PB ice flag
+flag_dur = calc_ice_duration(flag_ice_off, flag_ice_on)
 flag_error_dur = obs_dur - flag_dur
+    
+# store
+eval_df['ice_dur_avg_error'] = ice_dur_avg_error
+eval_df['ice_dur_rmse'] = ice_dur_rmse
+```
+
+# Add process-based ice flag evaluation to the dataframe
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.md:484]: After looping through all the models (retrieving measures of daily accuracy and date-based RMSEs), add the process-based results as a row

File: 03_eval/src/A_EvaluationMetrics.md [hunk 4]
```diff
@@ -266,149 +457,292 @@ def calc_ice_duration(off, on):
...
+ calc_rmse(flag_ice_on_error),
+ np.mean(flag_ice_off_error),
+ calc_rmse(flag_ice_off_error),
+ np.mean(flag_error_dur),
+ calc_rmse(flag_error_dur)]
+
+eval_df.loc[len(eval_df)] = new_row
+```
+
+# Find model that improved upon PB baseline the best (and second best)
```
Comment (jdiaz4302) [inline 03_eval/src/A_EvaluationMetrics.md:500]: Finding the "best" model(s) by finding those that reduce the average of date-based RMSEs the most

File: 03_eval/src/A_EvaluationMetrics.md [hunk 5]
```diff
@@ -266,149 +457,292 @@ def calc_ice_duration(off, on):
...
+# average percent change across the 3 columns
+eval_df['avg_rmse_change'] = np.mean(perc_change_rmse.values, axis = 1)
+```
+
+```python
+# rank all model realizations by their avg rmse improvement
+temp = eval_df['avg_rmse_change'].values.argsort()
+ranks = np.empty_like(temp)
+ranks[temp] = np.arange(len(eval_df['avg_rmse_change'].values))
+eval_df['avg_rmse_change_rank'] = ranks
```
Comment (AndyMcAliley) [inline 03_eval/src/A_EvaluationMetrics.md:523]: Ranking in Python came up in conversation with UMN recently. Bummer that this is the best way to do it in `numpy`. From [here](https://stackoverflow.com/questions/5284646/rank-items-in-an-array-using-python-numpy-without-sorting-array-twice) I see that `scipy.stats.rankdata()` ranks items, allowing different options for behavior for ties. I'll have to remember that function. Not saying you need to use it -it's nice to keep the number of package dependencies small.

**Review by AndyMcAliley — APPROVED** [review]
This all looks reasonable to me. I left some very minor comments and suggested changes, but they don't affect the functionality of the code.

File: 03_eval/src/A_EvaluationMetrics.md [hunk 6]
```diff
@@ -136,127 +242,212 @@ def calc_accuracy_iceflags(inputs, obs):
...
+# PB flag
+flag_ice_on_error = calc_date_errors(flag_ice_on, obs_ice_on)
+```
+
+```python
+# view distribution of errors
+plt.hist(flag_ice_on_error, label = 'process-based',
+         bins = range(-1*n_days_hist_cutoff, n_days_hist_cutoff), color = 'gray')
+plt.legend()
+plt.pause(0.001)
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 03_eval/src/A_EvaluationMetrics.md:368]: Every time you call `plt.pause()` matplotlib creates a new plot from the same cell. Here it separates the PB histograms from the ML histograms; I wanted the PB histogram to not be as hidden behind/with the ML ones.

If I loop 5 times and plot a line each time, there will be 5 lines on the same plot. If I do the same thing with a `plt.pause` at the end of each loop, there will be 5 plots with 1 line each.

File: 03_eval/src/A_EvaluationMetrics.md [hunk 7]
```diff
@@ -136,127 +242,212 @@ def calc_accuracy_iceflags(inputs, obs):
...
+        
+    return(ice_on_ids, ice_off_ids,
+           ice_on, ice_off)
+```
+
+```python
+# important dates = last ice on and last ice off
+# get important dates and date-indices for observations
+objects = extract_date_i_and_date(valid_y, valid_dates)
+obs_ice_on_ids, obs_ice_off_ids, obs_ice_on, obs_ice_off = objects
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 03_eval/src/A_EvaluationMetrics.md:303]: Yeah, exactly

File: 03_eval/src/A_EvaluationMetrics.md [hunk 8]
```diff
@@ -136,127 +242,212 @@ def calc_accuracy_iceflags(inputs, obs):
...
+    
+# PB flag
+flag_ice_on_error = calc_date_errors(flag_ice_on, obs_ice_on)
+```
+
+```python
+# view distribution of errors
+plt.hist(flag_ice_on_error, label = 'process-based',
+         bins = range(-1*n_days_hist_cutoff, n_days_hist_cutoff), color = 'gray')
+plt.legend()
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 03_eval/src/A_EvaluationMetrics.md:367]: Yeah, this is intended so that both of the two plots (separated by `plt.pause()`) will each have a legend

File: 03_eval/src/A_EvaluationMetrics.md [hunk 9]
```diff
@@ -266,149 +457,292 @@ def calc_ice_duration(off, on):
...
-print("\nModel duration RMSE:", calc_rmse(pred_error_dur))
+```python
+# get indices of date-based RMSEs in dataframe
+col_names_as_array = np.asarray(list(eval_df))
+i1 = np.argwhere(col_names_as_array == 'ice_on_rmse').item()
+i2 = np.argwhere(col_names_as_array == 'ice_off_rmse').item()
+i3 = np.argwhere(col_names_as_array == 'ice_dur_rmse').item()
+
+# get the percent change in RMSEs relative to PB ice flags
+perc_change_rmse = (eval_df.iloc[:, [i1, i2, i3]] - eval_df.iloc[-1, [i1, i2, i3]]) / eval_df.iloc[-1, [i1, i2, i3]]
```
[the remark this answers is not in the document]
Comment (jdiaz4302) [reply 03_eval/src/A_EvaluationMetrics.md:512]: Yeah, that's much better

File: 03_eval/src/A_EvaluationMetrics.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (AndyMcAliley) [reply 03_eval/src/A_EvaluationMetrics.md:368]: Interesting! I never knew that. Probably because I bite the bullet and explicitly create new figures using `fig, ax = plt.subplots()` when I plot multiple figures at once.
